### PR TITLE
[show][config] Add new snmp commands

### PR DESCRIPTION
**Pull request by tsvanduyn**
**- What I did**
I created these new show commands: 

show runningconfiguration snmp 
show runningconfiguration snmp contact
show runningconfiguration snmp location
show runningconfiguration snmp community
show runningconfiguration snmp user

show runningconfiguration snmp contact --json
show runningconfiguration snmp location --json
show runningconfiguration snmp community --json
show runningconfiguration snmp user --json 

These commands will check in the REDIS ConfigDB and get the information.

**- How I did it**
I created a standard snmp schema that was approved here: 
https://github.com/Azure/SONiC/pull/731

I also moved from using only the snmp.yml file to using the ConfigDB for the device. 
https://github.com/Azure/sonic-buildimage/pull/6205

Jinja template update: 
https://github.com/Azure/sonic-buildimage/pull/6134

This way the REDIS ConfigDB will be the source of truth but we can still use the snmp.yml file.

Since it is setup whenever the snmp docker container is restarted to pull this info into the ConfigDB we are able to use the data in the configdb for the show commands.
**- How to verify it**
You can run the show commands: 

show runningconfiguration snmp 
show runningconfiguration snmp contact
show runningconfiguration snmp location
show runningconfiguration snmp community
show runningconfiguration snmp user

show runningconfiguration snmp contact --json
show runningconfiguration snmp location --json
show runningconfiguration snmp community --json
show runningconfiguration snmp user --json 

and compare the output to what is in the REDIS ConfigDB.

```
admin@switch1:~$ redis-cli -n 4 keys "*" | grep -i snmp 
SNMP_USER|travis
SNMP_COMMUNITY|Joe
FEATURE|snmp
SNMP_COMMUNITY|public
SNMP|CONTACT
SNMP|LOCATION
SNMP_COMMUNITY|Travis
admin@switch1:~$ redis-cli -n 4 hgetall "SNMP_USER|travis"
 1) "SNMP_USER_ENCRYPTION_TYPE"
 2) "AES"
 3) "SNMP_USER_AUTH_TYPE"
 4) "SHA"
 5) "SNMP_USER_ENCRYPTION_PASSWORD"
 6) "TravEncryptPass"
 7) "SNMP_USER_AUTH_PASSWORD"
 8) "TravAuthPass"
 9) "SNMP_USER_TYPE"
10) "Priv"
11) "SNMP_USER_PERMISSION"
12) "RO"
admin@switch1:~$ redis-cli -n 4 hgetall "SNMP_COMMUNITY|Joe"
1) "TYPE"
2) "RO"
admin@switch1:~$ redis-cli -n 4 hgetall "SNMP|CONTACT"
1) "travis"
2) "travis@contoso.com"
admin@switch1:~$ redis-cli -n 4 hgetall "SNMP|LOCATION"
1) "Location"
2) "Palm Springs"
admin@switch1:~$
```

**- Previous command output (if the output of a command-line utility has changed)**
N/A
**- New command output (if the output of a command-line utility has changed)**
```
admin@switch1:~$ show run snmp
Location
------------
Palm Springs


SNMP_CONTACT    SNMP_CONTACT_EMAIL
--------------  --------------------
travis          travis@contoso.com


Community String    Community Type
------------------  ----------------
Joe                 RO
public              RO
Travis              RO


User    Type    Auth Type    Auth Password    Encryption Type    Encryption Password
------  ------  -----------  ---------------  -----------------  ---------------------
travis  Priv    SHA          TravAuthPass     AES                TravEncryptPass
admin@switch1:~$ show run snmp location 
Location
------------
Palm Springs
admin@switch1:~$ show run snmp contact
Contact    Contact Email
---------  ------------------
travis     travis@contoso.com
admin@switch1:~$ show run snmp community
Community String    Community Type
------------------  ----------------
Joe                 RO
public              RO
Travis              RO
admin@switch1:~$ show run snmp server
agentAddress udp:161
agentAddress udp6:161
admin@switch1:~$ show run snmp user
User    Type    Auth Type    Auth Password    Encryption Type    Encryption Password
------  ------  -----------  ---------------  -----------------  ---------------------
travis  Priv    SHA          TravAuthPass     AES                TravEncryptPass
admin@switch1:~$ show run snmp location --json
{'Location': 'Palm Springs'}
admin@switch1:~$ show run snmp contact --json
{'travis': 'travis@contoso.com'}
admin@switch1:~$ show run snmp community --json
{'Joe': {'TYPE': 'RO'}, 'public': {'TYPE': 'RO'}, 'Travis': {'TYPE': 'RO'}}
admin@switch1:~$ show run snmp user --json
{'travis': {'SNMP_USER_ENCRYPTION_TYPE': 'AES', 'SNMP_USER_AUTH_TYPE': 'SHA', 'SNMP_USER_ENCRYPTION_PASSWORD': 'TravEncryptPass', 'SNMP_USER_AUTH_PASSWORD': 'TravAuthPass', 'SNMP_USER_TYPE': 'Priv', 'SNMP_USER_PERMISSION': 'RO'}}
admin@switch1:~$
```

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+@snmp.group(cls=clicommon.AbbreviationGroup)
+@clicommon.pass_db
+def community(db):
+    pass
+
+
+def is_valid_community_type(commstr_type):
+    commstr_types = ['RO', 'RW']
+    if commstr_type not in commstr_types:
+        click.echo("Invalid community type.  Must be either RO or RW")
```
Comment (qiluo-msft) [inline config/main.py:2087]: >click.echo [](http://example.com/codeflow?start=8&length=10)

I see you have `click.echo` both inside and outside this checker function. The inner one seems redundant.

Comment (qiluo-msft) [reply config/main.py:2087]: You can just
```
return commstr_type in commstr_types
```

File: config/main.py [hunk 2]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        return False
+    return True
+
+
+def is_valid_user_type(user_type):
+    user_types = ['noauthnopriv', 'authnopriv', 'priv']
+    if user_type not in user_types:
+        click.echo("Invalid user type.  Must be one of these one of these three "
```
Comment (qiluo-msft) [inline config/main.py:2095]: >click.echo [](http://example.com/codeflow?start=8&length=10)

The same, and many more below.

File: config/main.py [hunk 3]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+
+@community.command('add')
+@click.argument('community', metavar='<snmp_community>', required=True)
+@click.argument('string_type', metavar='<RO|RW>', required=True)
+@clicommon.pass_db
+def add_community(db, community, string_type):
+    """ Add snmp community string"""
+    string_type = string_type.upper()
+    if not is_valid_community_type(string_type):
+        click.echo("SNMP type must be RO or RW and not {}".format(string_type))
```
Comment (qiluo-msft) [inline config/main.py:2187]: >echo [](http://example.com/codeflow?start=14&length=4)

If this is critical error message, you can echo to stderr instead of stdout.

File: config/main.py [hunk 4]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        sys.exit(1)
+    if not snmp_community_secret_check(community):
+        click.echo("Bad community")
```
Comment (qiluo-msft) [inline config/main.py:2190]: >Bad [](http://example.com/codeflow?start=20&length=3)

Better display the specific reason instread of `Bad`

File: config/main.py [hunk 5]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        sys.exit(2)
+    snmp_communities = db.cfgdb.get_table("SNMP_COMMUNITY")
+    if community in snmp_communities:
+        click.echo("SNMP community {} is already configured".format(community))
+        sys.exit(3)
+    else:
```
Comment (qiluo-msft) [inline config/main.py:2196]: >else: [](http://example.com/codeflow?start=4&length=5)

you don't need this `else:`, and can reduce one indentation  level of below section.

File: config/main.py [hunk 6]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        db.cfgdb.set_entry('SNMP_COMMUNITY', community, {'TYPE': string_type})
+        click.echo("SNMP community {} added to configuration".format(community))
+        try:
+            click.echo("Restarting SNMP service...")
+            clicommon.run_command("systemctl reset-failed snmp.service", display_cmd=False)
+            clicommon.run_command("systemctl restart snmp.service", display_cmd=False)
+        except SystemExit as e:
+            db.cfgdb.set_entry('SNMP_COMMUNITY', community, None)
```
Comment (qiluo-msft) [inline config/main.py:2204]: >set_entry [](http://example.com/codeflow?start=21&length=9)

Why you set SNMP_COMMUNITY again? It's dangerous to execute complex operation inside `except` block.

File: config/main.py [hunk 7]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        sys.exit(2)
+    elif new_community in snmp_communities:
+        click.echo("New SNMP community {} to replace current SNMP community {} already "
+                   "configured".format(new_community, current_community))
+        sys.exit(3)
+    else:
+        string_type = snmp_communities[current_community]['TYPE']
+        db.cfgdb.set_entry('SNMP_COMMUNITY', new_community, {'TYPE': string_type})
+        click.echo("SNMP community {} added to configuration".format(new_community))
+        db.cfgdb.set_entry('SNMP_COMMUNITY', current_community, None)
```
Comment (qiluo-msft) [inline config/main.py:2251]: >db.cfgdb.set_entry [](http://example.com/codeflow?start=8&length=18)

The purpose of `replace` CLI is to use single Redis transaction, but you uses 2 `set_entry`. Could you combine them into one?

File: config/main.py [hunk 8]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        click.echo('SNMP community {} removed from configuration'.format(current_community))
```
Comment (qiluo-msft) [inline config/main.py:2252]: >click.echo('SNMP community {} removed from configuration'.format(current_community)) [](http://example.com/codeflow?start=8&length=84)

Too verbose, you can remove this line and the next line is good enough.

File: config/main.py [hunk 9]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+            raise click.Abort()
+
+
+@snmp.group(cls=clicommon.AbbreviationGroup)
+@clicommon.pass_db
+def contact(db):
+    pass
+
+
+def is_valid_email(email):
```
Comment (qiluo-msft) [inline config/main.py:2269]: >is_valid_email [](http://example.com/codeflow?start=4&length=14)

Could you write some unit testcases for this function?

File: config/main.py [hunk 10]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+@click.argument('user_auth_password', metavar='<auth_password>', required=False)
+@click.argument('user_encrypt_type', metavar='<DES|AES>', required=False)
+@click.argument('user_encrypt_password', metavar='<encrypt_password>', required=False)
+@clicommon.pass_db
+def add_user(db, user, user_type, user_permission_type, user_auth_type, user_auth_password, user_encrypt_type, 
+             user_encrypt_password):
+    """ Add snmp user"""
+    click.echo(user)
+    if not snmp_username_check(user):
+        click.echo("Bad SNMP user")
```
Comment (qiluo-msft) [inline config/main.py:2494]: >Bad SNMP user [](http://example.com/codeflow?start=20&length=13)

This message is not friendly to user...

File: config/main.py [hunk 11]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+                    click.echo("User encrypt password is missing")
+                    sys.exit(12)
+                elif user_encrypt_password:
+                    if not snmp_user_secret_check(user_encrypt_password):
+                        click.echo("Bad user encrypt password")
+                        sys.exit(13)
+    snmp_users = db.cfgdb.get_table("SNMP_USER")
+    if user in snmp_users.keys():
+        click.echo("SNMP user {} is already configured".format(user))
+        sys.exit(14)
```
Comment (qiluo-msft) [inline config/main.py:2553]: >14 [](http://example.com/codeflow?start=17&length=2)

Better to define an Enum class to represent all the possible return codes.

File: show/main.py [hunk 12]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+            except KeyError:
+                snmp_user_type = 'Null'
+            snmp_user_body.append([snmp_user, snmp_user_permissions_type, snmp_user_type, snmp_user_auth_type, 
+                                   snmp_user_auth_password, snmp_user_encryption_type, snmp_user_encryption_password])
+        click.echo(tabulate(natsorted(snmp_user_body), snmp_user_header))
+
+
+# ("show runningconfiguration snmp")
+@clicommon.pass_db
+def show_run_snmp_global(db, ctx):
```
Comment (qiluo-msft) [inline show/main.py:1263]: >global [](http://example.com/codeflow?start=18&length=6)

What is the meaning of `global`?

File: show/main.py [hunk 13]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+@runningconfiguration.group("snmp", invoke_without_command=True)
+@clicommon.pass_db
+@click.pass_context
+def snmp(ctx, db):
+    """Show SNMP running configuration"""
+    if ctx.invoked_subcommand is None:
+       show_run_snmp_global(db.cfgdb)
+
+
+# '("show runningconfiguration snmp community")
```
Comment (qiluo-msft) [inline show/main.py:1133]: >'(" [](http://example.com/codeflow?start=2&length=3)

Could you remove the extra marks in comment?

File: show/main.py [hunk 14]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+        comm_string_type = snmp_comm_table[line]['TYPE']
+        snmp_comm_body.append([comm_string, comm_string_type])
+    click.echo(tabulate(natsorted(snmp_comm_body), snmp_comm_header))
+    click.echo("\n")
+    for line in snmp_users:
+        snmp_user = line
+        try:
+            snmp_user_permissions_type = snmp_users_table[line]['SNMP_USER_PERMISSION']
+        except KeyError:
+            snmp_user_permissions_type = 'Null'
```
Comment (qiluo-msft) [inline show/main.py:1306]: >'Null' [](http://example.com/codeflow?start=41&length=6)

Use a named constant?

File: tests/config_snmp_test.py [hunk 15]
```diff
@@ -0,0 +1,814 @@
...
+        db = Db()
+        runner = CliRunner()
+        with mock.patch('utilities_common.cli.run_command') as mock_run_command:
+            result = runner.invoke(config.config.commands["snmp"].commands["contact"].commands["add"],
+                                    ["testuser", "testuser@contoso.com"], obj=db)
+            print(result.exit_code)
+            print(result.output)
+            assert result.exit_code == 0
+            assert result.output == config_snmp_contact_add_del_new_contact
+            assert db.cfgdb.get_entry("SNMP", "CONTACT") == {"testuser": "testuser@contoso.com"}
```
Comment (SuvarnaMeenakshi) [inline tests/config_snmp_test.py:304]: same comment as above, is this add user required here?

**Review by SuvarnaMeenakshi — CHANGES_REQUESTED** [review]
As comments.

File: show/main.py [hunk 16]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+    """show SNMP running configuration community"""
+    snmp_comm_header = ["Community String", "Community Type"]
+    snmp_comm_body = []
+    snmp_comm_keys = db.cfgdb.get_table('SNMP_COMMUNITY')
+    snmp_comm_strings = snmp_comm_keys.keys()
+    if json_output:
+        try:
+            if snmp_comm_keys:
+                click.echo(snmp_comm_keys)
+        except KeyError:
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply show/main.py:1148]: I double checked this.   There shouldn't be a KeyError.  I can remove that if you want.  

I removed all the communities and tested: 
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages/config$ show run snmp community 
Community String    Community Type
------------------  ----------------
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages/config$ redis-cli -n 4 keys "*" | grep -i snmp 
SNMP_USER|travis
ACL_TABLE|SNMP_ACL
SNMP|LOCATION
FEATURE|snmp
SNMP|CONTACT

>>> config_db = ConfigDBConnector()
>>> config_db.connect()
>>> 
>>> config_db.get_table('SNMP_COMMUNITY')
{}
>>>

File: show/main.py (hunk 12 above)
Comment (tsvanduyn) [reply show/main.py:1263]: It has both the "contact" and the "location" and it is the root "SNMP" for the redis-db.  
If you'd like it named different that's fine.  Just let me know what you'd want.  
I also named it like this because "show run snmp" will print all out these SNMP commands 
show run snmp contact
show run snmp location
show run snmp community
show run snmp user
So I called it "global" because it encompassed all of these commands.

File: show/main.py (hunk 14 above)
Comment (tsvanduyn) [reply show/main.py:1306]: Are you saying you'd like me to have something like this? 
UNSET_USER_VAR = "Null"? 
and then in each except KeyError set the snmp_user parameter to that instead?

File: tests/config_snmp_test.py [hunk 17]
```diff
@@ -0,0 +1,814 @@
...
+        with mock.patch('utilities_common.cli.run_command') as mock_run_command:
+            result = runner.invoke(config.config.commands["snmp"].commands["community"].commands["replace"],
+                                                         ["Rainer", "Everest"], obj=db)
+        print(result.exit_code)
+        assert result.exit_code == 0
+        assert 'SNMP community Everest added to configuration' in result.output
+        assert db.cfgdb.get_entry("SNMP_COMMUNITY", "Everest") == \
+                expected_snmp_community_replace_existing_community_with_new_community_output
+
+    def test_config_snmp_community_replace_existing_community_does_not_exist(self):
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply tests/config_snmp_test.py:201]: Sure. I'll update that.

File: config/main.py (hunk 10 above)
Comment (tsvanduyn) [reply config/main.py:2494]: I didn't add more info there because the helper function "snmp_username_check" was providing that info above that statement.  
Per what you added above it looks like you'd prefer that I put all the logic in the "add_user" function and not the helper function.   
I think it makes more sense to put in in the helper function instead because that has the more specific info.  What do you think?

File: show/main.py (hunk 13 above)
Comment (tsvanduyn) [reply show/main.py:1133]: I'll remove the "'"

File: show/main.py [hunk 18]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+        snmp_global_table['LOCATION'] = ''
+    click.echo(tabulate(snmp_location_body, snmp_location_header))
+    click.echo("\n")
+    try:
+        if snmp_global_table['CONTACT']:
+            snmp_contact = list(snmp_global_table['CONTACT'].keys())
+            snmp_contact_email = [snmp_global_table['CONTACT'][snmp_contact[0]]]
+            snmp_contact_body.append([snmp_contact[0], snmp_contact_email[0]])
+    except KeyError:
+        snmp_global_table['CONTACT'] = ''
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply show/main.py:1290]: This is required.  If a contact is not set then we will get an error.  

admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$ show run snmp contact
Contact    Contact Email
---------  ---------------
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$ show run snmp contact --json
SNMP Contact not set
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$ 
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$ 
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$ show run snmp 
Location
----------


Traceback (most recent call last):
  File "/usr/local/bin/show", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1114, in invoke
    return Command.invoke(self, ctx)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/show/main.py", line 1216, in snmp
    show_run_snmp_global(db.cfgdb)
  File "/usr/local/lib/python3.7/dist-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/show/main.py", line 1367, in show_run_snmp_global
    if snmp_global_table['CONTACT']:
KeyError: 'CONTACT'
admin@str-s6000-acs-11:/usr/local/lib/python3.7/dist-packages$

File: show/main.py [hunk 19]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+                        "Encryption Password"]
+    snmp_user_body = []
+    if json_output:
+        try:
+            if snmp_users_keys:
+                click.echo(snmp_users_keys)
+        except KeyError:
+            click.echo("SNMP User not set")
+    else:
+        for line in snmp_users:
```
Comment (qiluo-msft) [inline show/main.py:1230]: >for line in snmp_users: [](http://example.com/codeflow?start=8&length=23)

You can iterate `snmp_users_keys` directly, get the dict key and value at the same time.

File: show/main.py [hunk 20]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+            snmp_user = line
+            try:
+                snmp_user_permissions_type = snmp_users_keys[line]['SNMP_USER_PERMISSION']
```
Comment (qiluo-msft) [inline show/main.py:1233]: >snmp_users_keys[line]['SNMP_USER_PERMISSION'] [](http://example.com/codeflow?start=45&length=45)

You can prevent try-except by `snmp_users_keys[line].get('SNMP_USER_PERMISSION', 'Null')`

File: show/main.py [hunk 21]
```diff
@@ -1133,6 +1120,215 @@ def ntp(verbose):
...
+
+
+# '("show runningconfiguration snmp user")
+@snmp.command('user')
+@click.option('--json', 'json_output', required=False, is_flag=True, type=click.BOOL, 
+              help="Display the output in JSON format")
+@clicommon.pass_db
+def users(db, json_output):
+    """show SNMP running configuration user"""
+    snmp_users_keys = db.cfgdb.get_table('SNMP_USER')
```
Comment (qiluo-msft) [inline show/main.py:1218]: >snmp_users_keys [](http://example.com/codeflow?start=4&length=15)

You get a table, why call it "keys"?

Comment (tsvanduyn) [reply show/main.py:1218]: That table will return a list of the SNMP_USERs.  I called it keys because I use user (basically acting as a key) to return the value of how that user is set.

Comment (tsvanduyn) [reply show/main.py:1218]: I looked at what you had closer and I'm updating this to just be snmp_users.

File: show/main.py (hunk 19 above)
Comment (tsvanduyn) [reply show/main.py:1230]: I've updated this to do what you've mentioned.

File: show/main.py (hunk 20 above)
Comment (tsvanduyn) [reply show/main.py:1233]: This is very cool.  I like this way better.  Thanks!

File: show/main.py (hunk 12 above)
Comment (tsvanduyn) [reply show/main.py:1263]: removed "Global" after looking at what you were saying.

File: show/main.py (hunk 14 above)
Comment (tsvanduyn) [reply show/main.py:1306]: Changed the way I was doing this per what you had above.  I don't think this is needed anymore.

File: config/main.py (hunk 6 above)
Comment (tsvanduyn) [reply config/main.py:2204]: If this failed when trying to set SNMP_COMMUNITY originally wouldn't it be better to try to remove it?  Or you would rather this just exit and not retry this?

File: config/main.py (hunk 7 above)
Comment (tsvanduyn) [reply config/main.py:2251]: I set this up to add a new one and then delete the old one after speaking with Joe when I originally wrote this.  As we thought that would be the safest way.

File: config/main.py (hunk 8 above)
Comment (tsvanduyn) [reply config/main.py:2252]: Sure I'm removing that.

File: config/main.py [hunk 22]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+        user_type = "Priv"
+    user_permission_type = user_permission_type.upper()
+    if not is_valid_community_type(user_permission_type):
+        click.echo("User permission type {} is invalid.  Must be RO or RW".format(user_permission_type))
+        sys.exit(3)
+    if user_type == "noAuthNoPriv":
+        if user_auth_type:
+            click.echo("User auth type not used with 'noAuthNoPriv'.  Please use 'AuthNoPriv' or 'Priv' instead")
+            sys.exit(4)
+    if user_type is not 'noAuthNoPriv':
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply config/main.py:2514]: I'm not sure what you mean here.  
I'm trying to account for cases where "noauthnopriv" is selected but the user also add an user_auth_type when it is unnecessary to give the user a hint on what they need.

File: config/main.py (hunk 6 above)
Comment (qiluo-msft) [reply config/main.py:2204]: I understand you are trying to recover. However set SNMP_COMMUNITY to None is not enough for a complete recovery, and it may confuse user when troubleshoot. Prefer exit directly.

File: config/main.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (SuvarnaMeenakshi) [reply config/main.py:2514]: What I meant is this:
Currently the code is:
if user_type == "noAuthNoPriv":
    ....
if user_type is not 'noAuthNoPriv':
   ...
What I meant to ask is, can the above 2 if checks be modified as:
if user_type == "noAuthNoPriv":
    ...
else:
   ...

File: tests/config_snmp_test.py [hunk 23]
```diff
@@ -0,0 +1,814 @@
...
+    def test_config_snmp_contact_add_with_existing_contact(self):
+        db = Db()
+        runner = CliRunner()
+        with mock.patch('utilities_common.cli.run_command') as mock_run_command:
+            result = runner.invoke(config.config.commands["snmp"].commands["contact"].commands["add"],
+                                    ["testuser", "testuser@contoso.com"], obj=db)
+            print(result.exit_code)
+            print(result.output)
+            assert result.exit_code == 0
+            assert result.output == config_snmp_contact_add_del_new_contact
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply tests/config_snmp_test.py:284]: So this is testing if there is an existing contact and what happens if we try to add a 2nd one.  The point is to not allow the 2nd one to be added. 
I have to add the original testuser contact first since we have no contact in the mock_table/config_db.json

File: tests/config_snmp_test.py (hunk 15 above)
Comment (tsvanduyn) [reply tests/config_snmp_test.py:304]: I've changed the function name to be more clear to this: 
test_config_snmp_contact_del_new_contact_when_contact_exists
I just want to test adding a deleting a contact.  
But I want there to be a contact already configured so I verify if there is a contact configure the del will fail because the wrong contact was trying to be deleted.

File: tests/config_snmp_test.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply tests/config_snmp_test.py:201]: updated.

File: tests/config_snmp_test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (SuvarnaMeenakshi) [reply tests/config_snmp_test.py:284]: Got the idea behind adding a contact first. The second contact cannot be added while first exists.

File: config/main.py (hunk 1 above)
Comment (tsvanduyn) [reply config/main.py:2087]: Using helper function for the comments per my discussion with Qi.

File: config/main.py (hunk 9 above)
Comment (tsvanduyn) [reply config/main.py:2266]: Added some additional tests for this.

File: tests/config_snmp_test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply tests/config_snmp_test.py:284]: Yep.   Thanks Suvarna.

File: show/main.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply show/main.py]: I've removed this.

File: config/main.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply config/main.py]: Updated to use "else"

File: config/main.py (hunk 11 above)
Comment (tsvanduyn) [reply config/main.py]: Updated to user Enum.

File: config/main.py [hunk 24]
```diff
@@ -2064,6 +2064,538 @@ def delete_snmptrap_server(ctx, ver):
...
+    user_type = user_type.lower()
+    if not is_valid_user_type(user_type):
+        click.echo("User type {} is invalid.  Must be noauthnopriv, authnopriv, or priv".format(user_type))
+        sys.exit(2)
+    if user_type == "noauthnopriv":
+        user_type = "noAuthNoPriv"
+    elif user_type == "authnopriv":
+        user_type = "AuthNoPriv"
+    elif user_type == "priv":
+        user_type = "Priv"
```
[the remark this answers is not in the document]
Comment (tsvanduyn) [reply config/main.py]: Updated this as well.

File: config/main.py (hunk 10 above)
Comment (tsvanduyn) [reply config/main.py]: removed this click.echo.

File: config/main.py (hunk 6 above)
Comment (tsvanduyn) [reply config/main.py]: ok removed.

File: config/main.py (hunk 5 above)
Comment (tsvanduyn) [reply config/main.py]: removed.

File: config/main.py (hunk 4 above)
Comment (tsvanduyn) [reply config/main.py]: removed this click.echo.  Using only in the helper function.

File: config/main.py [hunk 25]
```diff
@@ -2064,6 +2064,546 @@ def delete_snmptrap_server(ctx, ver):
...
+    if commstr_type not in commstr_types:
+        click.echo("Invalid community type.  Must be either RO or RW")
+        return False
+    return True
+
+
+def is_valid_user_type(user_type):
+    convert_user_type = {'noauthnopriv': 'noAuthNoPriv', 'authnopriv': 'AuthNoPriv', 'priv': 'Priv'}
+    user_types = ['noauthnopriv', 'authnopriv', 'priv']
+    if user_type not in user_types:
```
Comment (qiluo-msft) [inline config/main.py:2095]: >user_types [](http://example.com/codeflow?start=24&length=10)

You can directly use `convert_user_type`. No need to define `user_types`.

File: config/main.py (hunk 9 above)
Comment (qiluo-msft) [reply config/main.py:2266]: I see you added one test case `test_config_snmp_contact_add_invalid_email`. Actually you can test this function directly `is_valid_email` instead of using `CliRunner`.

Suggest add more test cases to cover as many as postive/negative for this regex.

File: config/main.py [hunk 26]
```diff
@@ -2064,6 +2064,546 @@ def delete_snmptrap_server(ctx, ver):
...
+@click.argument('user_encrypt_type', metavar='<DES|AES>', required=False)
+@click.argument('user_encrypt_password', metavar='<encrypt_password>', required=False)
+@clicommon.pass_db
+def add_user(db, user, user_type, user_permission_type, user_auth_type, user_auth_password, user_encrypt_type,
+             user_encrypt_password):
+    """ Add snmp user"""
+    if not snmp_username_check(user):
+        sys.exit(SnmpUserError.NameCheckFailure)
+    user_type = user_type.lower()
+    if not is_valid_user_type(user_type)[0]:
```
Comment (qiluo-msft) [inline config/main.py:2516]: >is_valid_user_type [](http://example.com/codeflow?start=11&length=18)

Better to call the function `is_valid_user_type` only once and store the return values into a tuple.

File: config/main.py (hunk 25 above)
Comment (tsvanduyn) [reply config/main.py]: fixed.

File: config/main.py (hunk 26 above)
Comment (tsvanduyn) [reply config/main.py]: updated.

File: config/main.py (hunk 25 above)
Comment (qiluo-msft) [reply config/main.py]: To further simplify, you can remove `.keys()`

**Review by qiluo-msft — APPROVED** [review]

File: config/main.py (hunk 9 above)
Comment (tsvanduyn) [reply config/main.py:2265]: added additional tests.

**Review by SuvarnaMeenakshi — APPROVED** [review]
